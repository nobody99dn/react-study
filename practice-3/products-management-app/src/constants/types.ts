export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum ProductTypes {
  Phone = 'Phone',
  Tablet = 'Tablet',
  Laptop = 'Laptop'
}

export const PRODUCT_TYPE_LIST: ProductTypes[] = [
  ProductTypes.Phone,
  ProductTypes.Tablet,
  ProductTypes.Laptop
];

export enum FilterOrderOptions {
  Asc = 'asc',
  Desc = 'desc'
}

export const ORDER_OPTIONS = [FilterOrderOptions.Asc, FilterOrderOptions.Desc];

export const localKey = 'localStates';
