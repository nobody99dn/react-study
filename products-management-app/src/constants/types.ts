export enum ButtonVariants {
  Default = 'default',
  Primary = 'primary',
  Secondary = 'secondary'
}

export enum FwType {
  Bold = 700,
  Normal = 400,
  Light = 300
}

export enum TextFieldVariants {
  Standard = 'standard',
  Outline = 'outline',
  Error = 'error'
}

export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum FormVariants {
  Create = 'Create',
  Edit = 'Edit'
}

export const PRODUCT_TYPES: string[] = [ 'Phone', 'Tablet', 'Laptop']

