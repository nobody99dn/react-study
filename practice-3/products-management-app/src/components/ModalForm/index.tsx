// Components
import Form from '@components/Form';
import Modal from '@components/Modal';

// Library
import React, { FormEvent } from 'react';

// Constants
import { FormVariants, PRODUCT_TYPE_LIST } from '@constants/index';

// Model
import { Product } from '@models/product';

export interface ModalFormProps {
  product: Product;
  isModalShow: boolean;
  validateMessage: string;
  isButtonLoading: boolean;
  handleSubmitForm: (event: FormEvent, product: Product) => void;
  handleCloseModal: () => void;
}

const ModalForm: React.FC<ModalFormProps> = ({
  product,
  isModalShow,
  validateMessage = '',
  isButtonLoading,
  handleSubmitForm,
  handleCloseModal
}) => (
  <Modal isVisible={isModalShow} handleClose={handleCloseModal}>
    <Form
      variants={!product.id ? FormVariants.Create : FormVariants.Edit}
      options={PRODUCT_TYPE_LIST}
      productItem={product}
      validateMessage={validateMessage}
      isButtonLoading={isButtonLoading}
      handleSubmit={handleSubmitForm}
    />
  </Modal>
);

export default ModalForm;
