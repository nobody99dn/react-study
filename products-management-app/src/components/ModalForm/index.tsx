// Components
import Form from '@components/Form';
import Modal from '@components/Modal';

// Library
import React, { FormEvent } from 'react';

// Constants
import { FormVariants, PRODUCT_TYPE_LIST } from '@constants/types';

// Type
import { Product } from '@models/product';

interface ModalFormProps {
  product: Product;
  isModalShow: boolean;
  validateMessage: string;
  isButtonLoading: boolean;
  handleSubmitForm: (event: FormEvent, product: Product) => void;
  handleCloseModal: () => void;
}

const ModalForm: React.FC<ModalFormProps> = ({
  product,
  isModalShow,
  validateMessage = '',
  isButtonLoading,
  handleSubmitForm,
  handleCloseModal
}) => (
  <Modal
    isVisible={isModalShow}
    onClose={handleCloseModal}
  >
    <Form
      variants={!product.id ? FormVariants.Create : FormVariants.Edit}
      options={PRODUCT_TYPE_LIST}
      productItem={product}
      validateMessage={validateMessage}
      isButtonLoading={isButtonLoading}
      onSubmit={handleSubmitForm}
    />
  </Modal>
);

export default ModalForm;
