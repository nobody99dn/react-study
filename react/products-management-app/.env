VITE_BASE_URL=https://62943cf0a7203b3ed0650c9b.mockapi.io
