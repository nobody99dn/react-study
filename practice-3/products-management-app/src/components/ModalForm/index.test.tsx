// Libraries
import { render, act, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import ModalForm, { ModalFormProps } from './index';
import { ProductTypes } from '@constants/types';

describe('ModalForm component', () => {
  const defaultProps: ModalFormProps = {
    product: {
      id: '1',
      imageUrl: '',
      name: 'Product test',
      price: 1000000,
      type: ProductTypes.Phone
    },
    isModalShow: true,
    validateMessage: 'test message',
    isButtonLoading: false,
    handleCloseModal: jest.fn(),
    handleSubmitForm: jest.fn()
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<ModalForm {...defaultProps} />).toJSON();
    expect(defaultTree).toMatchSnapshot();
  });

  test('should render correct validate message', () => {
    const message: string = 'Test message';

    const { getByText } = render(
      <ModalForm {...defaultProps} validateMessage={message} />,
      { container }
    );

    expect(getByText(message)).toBeInTheDocument();
  });

  test('should be click actions', () => {
    const submitMock = jest.fn();
    const closeMock = jest.fn();

    render(
      <ModalForm
        {...defaultProps}
        handleCloseModal={closeMock}
        handleSubmitForm={submitMock}
      />
    ),
      { container };

    const img: HTMLImageElement = screen.getByRole('img');

    const form: HTMLImageElement = screen.getByRole('form');

    fireEvent.click(img);

    fireEvent.submit(form);

    expect(closeMock).toHaveBeenCalled();
    expect(submitMock).toHaveBeenCalled();
  });
});
