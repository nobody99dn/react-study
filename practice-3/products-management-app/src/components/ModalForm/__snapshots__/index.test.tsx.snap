// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalForm component should render correctly 1`] = `
<div
  className="modal"
>
  <div
    className="modal-dialog"
  >
    <img
      alt="close icon"
      className="image  close-icon"
      onClick={[MockFunction]}
      onError={[Function]}
      onLoad={[Function]}
    />
    <div
      className="modal-body"
    >
      <div
        className="form-wrapper"
      >
        <div
          className="title"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
              "fontStyle": "inherit",
              "fontWeight": undefined,
              "padding": undefined,
            }
          }
        >
          Edit
           Product
        </div>
        <form
          aria-label="form"
          className="form"
          onSubmit={[Function]}
        >
          <fieldset
            className="field-group"
          >
            <div
              className="field-wrapper"
            >
              <label>
                Product name
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  name="name"
                  onChange={[Function]}
                  placeholder="Enter product name..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="text"
                  value="Product test"
                />
              </div>
            </div>
            <div
              className="select-wrapper"
            >
              <label>
                Product type
                :
              </label>
              <select
                className="select"
                disabled={false}
                name="type"
                onChange={[Function]}
                value="Phone"
              >
                <option
                  value=""
                >
                  Select
                </option>
                <option
                  value="Phone"
                >
                  Phone
                </option>
                <option
                  value="Tablet"
                >
                  Tablet
                </option>
                <option
                  value="Laptop"
                >
                  Laptop
                </option>
              </select>
            </div>
            <div
              className="field-wrapper"
            >
              <label>
                Price
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  name="price"
                  onChange={[Function]}
                  placeholder="Enter price..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="number"
                  value={1000000}
                />
              </div>
            </div>
            <div
              className="field-wrapper"
            >
              <label>
                Image link
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  name="imageUrl"
                  onChange={[Function]}
                  placeholder="Enter image link..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="text"
                  value=""
                />
              </div>
            </div>
            <p
              className="text-highlight"
              style={
                Object {
                  "color": "var(--danger)",
                  "fontSize": undefined,
                  "fontStyle": "inherit",
                  "fontWeight": 400,
                }
              }
            >
              test message
            </p>
          </fieldset>
          <button
            className="btn btn-primary"
            disabled={false}
          >
            Edit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
